File: server/app/admin/service.py
import hmac
import secrets
from datetime import datetime, timedelta, timezone

from fastapi import HTTPException, Request, Response, status

from app.auth.service import SESSION_COOKIE
from app.shared.config import get_settings
from app.shared.database import get_db

ADMIN_COOKIE = "gallery_admin_session"
ADMIN_SESSION_HOURS = 12


def require_admin(request: Request) -> dict:
    user_admin = admin_from_user_session(request)
    if user_admin:
        return user_admin
    token = request.cookies.get(ADMIN_COOKIE)
    if not token:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="请先登录管理员")
    with get_db() as db:
        row = db.execute(
            """
            SELECT token, expires_at
            FROM admin_sessions
            WHERE token = ? AND expires_at > ?
            """,
            (token, datetime.now(timezone.utc).isoformat()),
        ).fetchone()
    if not row:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="管理员登录已过期")
    return {"role": "admin", "source": "password"}


def admin_from_user_session(request: Request) -> dict | None:
    token = request.cookies.get(SESSION_COOKIE)
    if not token:
        return None
    with get_db() as db:
        row = db.execute(
            """
            SELECT users.id, users.username
            FROM sessions
            JOIN users ON users.id = sessions.user_id
            WHERE sessions.token = ?
              AND sessions.expires_at > ?
              AND users.is_admin = 1
            """,
            (token, datetime.now(timezone.utc).isoformat()),
        ).fetchone()
    if not row:
        return None
    return {"role": "admin", "source": "user", "user_id": row["id"], "username": row["username"]}


def login_admin(password: str, response: Response) -> None:
    settings = get_settings()
    if not hmac.compare_digest(password, settings.admin_password):
        raise HTTPException(status_code=401, detail="管理员密码错误")
    token = secrets.token_urlsafe(32)
    expires = datetime.now(timezone.utc) + timedelta(hours=ADMIN_SESSION_HOURS)
    with get_db() as db:
        db.execute(
            "INSERT INTO admin_sessions (token, expires_at) VALUES (?, ?)",
            (token, expires.isoformat()),
        )
    response.set_cookie(
        ADMIN_COOKIE,
        token,
        max_age=ADMIN_SESSION_HOURS * 3600,
        httponly=True,
        samesite="lax",
    )


def logout_admin(request: Request, response: Response) -> None:
    token = request.cookies.get(ADMIN_COOKIE)
    if token:
        with get_db() as db:
            db.execute("DELETE FROM admin_sessions WHERE token = ?", (token,))
    response.delete_cookie(ADMIN_COOKIE)

File: server-go/internal/admin/service.go
package admin

import (
	"crypto/hmac"
	"crypto/rand"
	"database/sql"
	"encoding/base64"
	"errors"
	"net/http"
	"time"

	"github.com/QWiseDev/Img2Gallery/server-go/internal/auth"
	"github.com/QWiseDev/Img2Gallery/server-go/internal/config"
)

const (
	CookieName        = "gallery_admin_session"
	adminSessionHours = 12
)

type Service struct {
	db   *sql.DB
	cfg  config.Config
	auth *auth.Service
}

func NewService(database *sql.DB, cfg config.Config, authService *auth.Service) *Service {
	return &Service{db: database, cfg: cfg, auth: authService}
}

func (s *Service) Login(password string) (string, int, error) {
	if !hmac.Equal([]byte(password), []byte(s.cfg.AdminPassword)) {
		return "", 0, ErrBadAdminPassword
	}
	token := token(32)
	expires := time.Now().UTC().Add(adminSessionHours * time.Hour)
	_, err := s.db.Exec("INSERT INTO admin_sessions (token, expires_at) VALUES (?, ?)", token, utcISO(expires))
	return token, adminSessionHours * 3600, err
}

func (s *Service) Logout(token string) error {
	if token == "" {
		return nil
	}
	_, err := s.db.Exec("DELETE FROM admin_sessions WHERE token = ?", token)
	return err
}

func (s *Service) Require(r *http.Request) (map[string]any, error) {
	if adminUser, ok := s.adminFromUser(r); ok {
		return adminUser, nil
	}
	cookie, _ := r.Cookie(CookieName)
	if cookie == nil {
		return nil, ErrAdminLoginRequired
	}
	var stored string
	err := s.db.QueryRow("SELECT token FROM admin_sessions WHERE token = ? AND expires_at > ?", cookie.Value, utcISO(time.Now().UTC())).Scan(&stored)
	if err != nil {
		return nil, ErrAdminSessionExpired
	}
	return map[string]any{"role": "admin", "source": "password"}, nil
}

func (s *Service) adminFromUser(r *http.Request) (map[string]any, bool) {
	cookie, _ := r.Cookie(auth.SessionCookie)
	if cookie == nil {
		return nil, false
	}
	user, err := s.auth.CurrentUser(cookie.Value)
	if err != nil || !user.IsAdmin {
		return nil, false
	}
	return map[string]any{
		"role": "admin", "source": "user", "user_id": user.ID, "username": user.Username,
	}, true
}

func token(size int) string {
	buf := make([]byte, size)
	if _, err := rand.Read(buf); err != nil {
		return base64.RawURLEncoding.EncodeToString([]byte(time.Now().String()))
	}
	return base64.RawURLEncoding.EncodeToString(buf)
}

func utcISO(value time.Time) string {
	return value.UTC().Format("2006-01-02T15:04:05.000000+00:00")
}

var (
	ErrBadAdminPassword    = errors.New("bad admin password")
	ErrAdminLoginRequired  = errors.New("admin login required")
	ErrAdminSessionExpired = errors.New("admin session expired")
)
